spark/src/main/scala/org/apache/comet/serde/arithmetic.scala

```diff
@@ -0,0 +1,293 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.comet.serde
+
+import scala.math.min
+
+import org.apache.spark.sql.catalyst.expressions.{Add, Attribute, Divide, EqualTo, EvalMode, Expression, If, IntegralDivide, Literal, Multiply, Remainder, Subtract}
+import org.apache.spark.sql.types.{BinaryType, BooleanType, ByteType, DataType, DateType, DecimalType, DoubleType, FloatType, IntegerType, LongType, NullType, ShortType, StringType, TimestampNTZType, TimestampType}
+
+import org.apache.comet.CometSparkSessionExtensions.withInfo
+import org.apache.comet.expressions.CometEvalMode
+import org.apache.comet.serde.QueryPlanSerde.{castToProto, evalModeToProto, exprToProtoInternal, serializeDataType}
+import org.apache.comet.shims.CometEvalModeUtil
+
+trait MathBase {
+  def createMathExpression(
+      expr: Expression,
+      left: Expression,
+      right: Expression,
+      inputs: Seq[Attribute],
+      binding: Boolean,
+      dataType: DataType,
+      evalMode: EvalMode.Value,
+      f: (ExprOuterClass.Expr.Builder, ExprOuterClass.MathExpr) => ExprOuterClass.Expr.Builder)
+      : Option[ExprOuterClass.Expr] = {
+    val leftExpr = exprToProtoInternal(left, inputs, binding)
+    val rightExpr = exprToProtoInternal(right, inputs, binding)
+
+    if (leftExpr.isDefined && rightExpr.isDefined) {
+      // create the generic MathExpr message
+      val builder = ExprOuterClass.MathExpr.newBuilder()
+      builder.setLeft(leftExpr.get)
+      builder.setRight(rightExpr.get)
+      builder.setEvalMode(evalModeToProto(CometEvalModeUtil.fromSparkEvalMode(evalMode)))
+      serializeDataType(dataType).foreach { t =>
+        builder.setReturnType(t)
+      }
+      val inner = builder.build()
+      // call the user-supplied function to wrap MathExpr in a top-level Expr
+      // such as Expr.Add or Expr.Divide
+      Some(
+        f(
+          ExprOuterClass.Expr
+            .newBuilder(),
+          inner).build())
+    } else {
+      withInfo(expr, left, right)
+      None
+    }
+  }
+
+  def nullIfWhenPrimitive(expression: Expression): Expression =
+    if (isPrimitive(expression)) {
+      val zero = Literal.default(expression.dataType)
+      expression match {
+        case _: Literal if expression != zero => expression
+        case _ =>
+          If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+      }
+    } else {
+      expression
+    }
+
+  private def isPrimitive(expression: Expression): Boolean = expression.dataType match {
+    case _: ByteType | _: ShortType | _: IntegerType | _: LongType | _: FloatType |
+        _: DoubleType | _: TimestampType | _: DateType | _: BooleanType | _: DecimalType =>
+      true
+    case _ => false
+  }
```

Comment:
This is moved without modification from QueryPlanSerde. I am not sure why this method is needed. I'm pretty sure that we can only perform arithmetic on primitive types?

Reply:
I'm not sure why this is needed either, but while we are checking, should this also check for TimestampNTZ?

Reply:
Spark only supports numeric and interval types for the arithmetic expressions, so I have updated the `supportedDataType` method in this PR to match that by removing types such as strings, timestamps, and dates. I have also removed `isPrimitive` now since only primitive types are supported.